### PR TITLE
24.11 rename codeName to Vicuna

### PR DESCRIPTION
**Pull request by sg-qwt**
1. We need to keep code name simple and universal, not every culture know how to type ñ on their keyboard
2. Avoid unnecessary and potential trouble caused by unicode 
https://github.com/NixOS/nixpkgs/issues/315574  
3. According to wiki, vicuña is same as vicuna

>  The vicuña or vicuna is one of the two wild South American camelids, which live in the high alpine areas of the Andes, the other being the guanaco, which lives at lower elevations. [Wikipedia](https://en.wikipedia.org/wiki/Vicu%C3%B1a)

### CODE REVIEW

**Review by SuperSandro2000 — APPROVED** [review]
We don't need to make our life's harder than they need to be and find more such bugs in other ecosystems.

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Until proven wrong, I do not think any of the bug related to mishandling of unicode has been a showstopper.

Release managers should be involved in this decision.

**Review by K900 — CHANGES_REQUESTED** [review]
We should, in fact, make our lives harder and find bugs in other ecosystems, as non-ASCII handling is important for a lot of users. We can revisit this if there are outstanding major blockers before release.

**Comment by RaitoBezarius** [comment]
> We need to keep code name simple and universal, not every culture know how to type ñ on their keyboard

I would argue this is because *we* are universal that we need to prepare for Unicode characters, this is about celebrating all the folks coming to use our distribution.

**Review by samueldr — CHANGES_REQUESTED** [review]
We shouldn't.

* * *

The only citation for *Vicuna* as such in the Wikipedia article is:

> The spelling vicuña is not even mentioned in some dictionaries, for example the [Macmillan Dictionary](https://www.macmillandictionary.com/dictionary/british/vicuna?q=vicu%C3%B1a).

With the broken link as shared. No web archive of the page exists.

The deficiencies of some arbitrary English language dictionaries shouldn't be used as an excuse to not use the correct name for the thing. 

* * *

Now, about bugs from elsewhere. The lack of preparation from other projects for valid text strings as arbitrary input is not our bug.

This is not only valid Unicode text (encoded as UTF-8), the character isn't even "special" in many ways since it was even in [CP437](https://en.wikipedia.org/wiki/Code_page_437).

Here it is in DOS 4.01:

image

If some software is less well internationalized and localizable than an IBM PC with a DOS from the mid 80s, the problem is not NixOS.

**Comment by mweinelt** [comment]
> We need to keep code name simple and universal, not every culture know how to type ñ on their keyboard

The animals don't carry any relevant meaning. Just refer to the releases by their date versioning.

> Avoid unnecessary and potential trouble caused by unicode

Fixing Unicode compatibility throughout the ecosystem is a worth goal, and it is beneficial for the .NET ecosystem as well!

At this point, I don't believe we have a strong motivation to accept this change. We can revisit this topic, if and when we do. Until then, let's keep fixing these kinds of issues.

**Comment by SuperSandro2000** [comment]
> Until proven wrong, I do not think any of the bug related to mishandling of unicode has been a showstopper.

The linked above dotnet cli is a real show stopper and until the library update is propagated to all tools, you cannot use it and the update scripts are not working. For my understanding, that is a show stopper.

You can reproduce that with:
- nix-build -A ArchiSteamFarm.fetch-deps
- ./result

That will fail with ``error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json`` regardless if ``AZURE_TELEMETRY_DISABLED=1`` is set or not.

**Comment by SuperSandro2000** [comment]
> We need to keep code name simple and universal, not every culture know how to type ñ on their keyboard

After doing some research I finally figured out how to type it on my keyboard.
You just need to hold Ctrl+Shift and then u, type 00F1 and press enter.

**Comment by hraban** [comment]
> > We need to keep code name simple and universal, not every culture know how to type ñ on their keyboard
> 
> After doing some research I finally figured out how to type it on my keyboard. You just need to hold Ctrl+Shift and then u, type 00F1 and press enter.

If you're into so-called [dead keys](https://en.wikipedia.org/wiki/Dead_key) it's <kbd>~</kbd><kbd>n</kbd>

**Comment by bobvanderlinden** [comment]
> We don't need to make our life's harder than they need to be and find more such bugs in other ecosystems.

I agree.

I'd rather not try to force the world to adhere to NixOS naming convention. The name is just for fun. It holds very little technical value. I would be really sad if it becomes a blocker or source of frustration for newcomers to adopt NixOS.

**Comment by Aleksanaa** [comment]
Needing further discussion.

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Release 24.11 (“Vicuña”, 2024.11/??) {#sec-release-24.11}
+# Release 24.11 (“Vicuna”, 2024.11/??) {#sec-release-24.11}
```
Comment (Mic92) [inline nixos/doc/manual/release-notes/rl-2411.section.md:1]:
```suggestion
# Release 24.11 (“Vicuña”, 2024.11/??) {#sec-release-24.11}
```

Comment (Mic92) [reply nixos/doc/manual/release-notes/rl-2411.section.md:1]: Let's keep it here for vanity reasons.

**Review by pluiedev — CHANGES_REQUESTED** [review]
IMO the argument that Vicuna is equivalent to Vicuña is like saying pinata is an accepted alternate spelling for piñata, or papier-mache is an OK way to spell papier-mâché — it's only okay ***in English*** because English words don't generally use diacritics and so it's deemed acceptable to render a word in another language that does use diacritics without diacritics. I find this approach overly Anglocentric.

On the technical side of things, yes, it sucks that it broke .NET, yes, it sucks that my BIOS firmware uses a font that renders it like Vicu�a. But had we never picked a codename that went beyond the bounds of 7-bit ASCII, that bug would have never been caught, due to the systemtic Anglocentric bias in tech itself — everyone uses English, so why bother testing anything beyond English? In that sense, the people that have to change are them, not us. 

And ñ is just part of the basic set of characters that fits within extended ASCII! Imagine what would happen had we picked something even more exotic...

**Comment by Mic92** [comment]
I had a chat with @wegank and we both agreed, that we should merge this change as otherwise it feels like being hostile to a certain part of the community that relies on this software.
The dotnet community seems to have done their part and fixed the issue: https://github.com/Azure/azure-sdk-for-net/pull/44386 but this won't fix old nixpkgs versions or the situation in short term. This would also render old binary cache generations that we keep around while paying a significant amount of money, useless.

**Comment by K900** [comment]
Have we seen any evidence of breakage other than that one .NET bug that's already fixed?

**Comment by Mic92** [comment]
> Have we seen any evidence of breakage other than that one .NET bug that's already fixed?

I am not aware of this being fixed in nixpkgs itself, it's fixed in some upstream library.
Apparently also firmware, also there it's seems more of a cosmetic issue.

**Comment by pluiedev** [comment]
> I am not aware of this being fixed in nixpkgs itself, it's fixed in some upstream library. Apparently also firmware, also there it's seems more of a cosmetic issue.

It's very much a cosmetic issue and doesn't at all impact functionality — a minor annoyance at most :+1:

**Comment by mweinelt** [comment]
@nazarewk Can you update us on the dotnet situation? Are we in a position to update/patch, and if not what is the timeline for that?

**Comment by Mic92** [comment]
Ok. Fixing this issue is nice, but I don't think we should burn out our dotnet contributors over this. If we can fix the issue, we can restore the old codename.

**Comment by SuperSandro2000** [comment]
> The dotnet community seems to have done their part and fixed the issue: [Azure/azure-sdk-for-net#44386](https://github.com/Azure/azure-sdk-for-net/pull/44386) but this won't fix old nixpkgs versions or the situation in short term.

This is a bug in the linked library *and* the dotnet-cli which isn't fixed yet and when fixed, maybe needs to be backported to 3 major versions.

@pluiedev my last name contains an ä which is part of the standard german character set everyone knows and is on all standard keyboards and despite that my german employer is not able to use that everywhere and I am very well accustomed that it gets replaced by ae.

Also I own a computer which I bought in Germany with by default german language and that isn't even able to display umlauts properly. 

It's sad but we can't choose every fight and uphill battle and you need to choose which ones to take.

**Comment by SuperSandro2000** [comment]
Let's take a look 11 years back onto LWN https://lwn.net/Articles/544794/ where Fedora tried something similar:

> But, to others, simply changing the release name amounts to "papering over" the real issue, which is ensuring that the build and QA tools can handle arbitrary UTF-8 text. Surely it is better to spend a little time now to fix the issues than to avoid them, the thinking went. Williamson, however, disagreed, [calling it](https://lwn.net/Articles/545769/) "a question of priorities" in light of Fedora's human resources and release schedule. Later, he [elaborated](https://lwn.net/Articles/545770/) that fixing UTF-8 support in the problematic tools in separate branches would be acceptable, if it did not slow down the release:

> You want to set up a side project to spin some images with crazy release names and see what breaks and fix that, then you know, go for it. But I'm trying to ship an operating system that works here, and leaving something we know is causing all kinds of problems in the problematic state just so we can keep finding exciting new problems to fix does not suffuse me with joy.
If we have to compromise on just papering it over for Alpha, I mean, _fine_. But seriously: sometimes papering it over is just the right thing to do.

**Comment by RaitoBezarius** [comment]
11 years later, we still do not have the backbone to unite against this sort of issues. A real loss for the celebration of diversity and upholding our values of welcoming everyone.

This is disappointing to say the least.

**Comment by pluiedev** [comment]
History truly repeats itself, huh... But still, it's the year 2024 — there isn't really an excuse to not support UTF-8 except due to legacy cruft that nobody has touched in years. I'd argue that it's a better time now to truly fix things that otherwise would have never been fixed.

**Comment by SuperSandro2000** [comment]
Then please support us by helping to fix the issues, instead of neglecting them. The dotnet issue is this one https://github.com/NixOS/nixpkgs/issues/315574 and the hot bread crumbs I summarized here https://github.com/NixOS/nixpkgs/issues/315574#issuecomment-2157012978 . Upstream issues are https://github.com/dotnet/runtime/issues/102819 and https://github.com/NuGet/Home/issues/13531

**Comment by bobvanderlinden** [comment]
> 11 years later, we still do not have the backbone to unite against this sort of issues. A real loss for the celebration of diversity and upholding our values of welcoming everyone.
> 
> This is disappointing to say the least.

Regardless of the change in NixOS, you can still fix the issues mentioned in this thread :sweat_smile:

**Comment by purepani** [comment]
> 11 years later, we still do not have the backbone to unite against this sort of issues. A real loss for the celebration of diversity and upholding our values of welcoming everyone.
> 
> This is disappointing to say the least.

As one of the many people who spent a very non-negligible time trying to figure out why on earth I couldn't use the dotnet cli at all(days before the issue was posted iirc), and then spent more time trying to come up with somewhat decent solutions to this, and given how much stuff this breaks, it is probably better to not make every single dotnet contributor manually change their OS name(which they might not even have the correct permissions to do in their environment) until there is some way to work around it. 

This is a temporary one character change that lets people actually use the dotnet cli, and to use the nix tooling to update and modify the entire dotnet ecosystem in nixpkgs. I don't find it super reasonable to force so many people to have a nonfunctional development environment if there hasn't been a fix in over 2 weeks.

**Comment by infinisil** [comment]
Iirc there's a word for this (grey/brown-out?), where you *temporarily* break something just to see what breaks downstream, but then revert it later, kind of like a warning shot. Doing this gives people downstream a clear signal that they really need to migrate/fix their stuff but also giving them time to do it. This approach is useful when it would be impossible to inform people of upcoming changes otherwise. Let's treat it as such a case.

> A real loss for the celebration of diversity and upholding our values of welcoming everyone.

That's a misleading way of framing this issue. From what I've seen, *everybody* here (and downstream!) wants to have good Unicode support all around, nobody is arguing against that. Instead, the reason for this change is because it's the best way to fix a problem for many users.

And from what I've seen, the only people not feeling welcome here right now are the ones that depend on software that can't handle Vicuña. But if there's anybody from e.g. Peru, I'd love to hear your thoughts. At least personally, I'm totally fine with everybody writing my home town as Zurich instead of Zürich :stuck_out_tongue:

**Comment by pluiedev** [comment]
I agree with the change being a pragmatic, temporary measure to unbreak .NET, but I disagree with this being a permanent change like previously implied — after all, there's no reason to change back if both forms are the same, right? (They're not)

When the dust settles and upstream implements measures to support the ñ properly, let's roll it back.

**Comment by RaitoBezarius** [comment]
> > A real loss for the celebration of diversity and upholding our values of welcoming everyone.
> 
> That's a misleading way of framing this issue. From what I've seen, _everybody_ here (and downstream!) wants to have good Unicode support all around, nobody is arguing against that. Instead, the reason for this change is because it's the best way to fix a problem for many users.
> 
> And from what I've seen, the only people not feeling welcome here right now are the ones that depend on software that can't handle Vicuña. But if there's anybody from e.g. Peru, I'd love to hear your thoughts. At least personally, I'm totally fine with everybody writing my home town as Zurich instead of Zürich 😛

That's also a misleading way of framing the issue. Patching exist.

If it's a source build, we can patch the sources.
If it's a binary, we can patch the binary.

This is just a matter of will and caring enough about the problem space. What makes it hard is when everyone has already concluded there's nothing left to do, when upstream reacted quite fast and provided precise and detailed guidance on how to fix the situation.

Thank you for clarifying your position.

**Comment by purepani** [comment]
> What makes it hard is when everyone has already concluded there's nothing left to do
Nobody has said that there's nothing left to do. This is a simple solution that we can use to make dotnet work, while we wait for a fix upstream or for a patch.

That being said, in the small chance than neither of those 2 things happen when it gets closer to the 24.11 release, perhaps it would be a good idea to make the original issue a blocker for release until upstream is fixed, or a patch is submitted. In the unlikely case that it's a lot closer to release without a solution, then maybe the discussion should happen again about the name.

**Comment by infinisil** [comment]
Correct me if I'm wrong, but updating to Vicuña breaks _running_ any package using the popular Azure Core .NET library, including third-party packages that we have no control over, even ones that might've ran perfectly fine for years without changes. So even if we had patches for Nixpkgs (which we don't yet, and it [might be tricky](https://github.com/NixOS/nixpkgs/issues/315574#issuecomment-2158886489) and take a while), _every_ use site would have to update their Nixpkgs pin to fix it too.

On the other hand, we can save this trouble for so many users with this inconsequential _one-character_ change on our side. Yes, there's multiple solutions (there always is), but there's also tradeoffs involved, and in this case they clearly favor this one imo.

And if somebody really wants to have a meaningful impact on language diversity, I recommend investing that energy into localisation support for our documentation and error messages!

**Comment by purepani** [comment]
Yes, it breaks running anything that dynamically reads `/etc/os-release` and reads the `PRETTY_NAME` field in that file if that program doesn't handle non-ascii characters correctly.
This includes anything using Azure.Core, which the original issue was about, and any version of the dotnet cli released over the past year when the runtime changed `RuntimeInfo.OSDescription` to read from the  `PRETTY_NAME`(which was done due to [this issue](https://github.com/dotnet/runtime/issues/83287).

**Comment by zimbatm** [comment]
I laughed so hard. It's such a typical NixOS thread. It's a good thing it was only a single character to change🤣

The best was this:
> 11 years later, we still do not have the backbone to unite against this sort of issue. A real loss for the celebration of diversity and upholding our values of welcoming everyone.

Brothers and sisters! We stand united behind this single ñ character! To celebrate our diversity of characters and uphold our UTF-8 code points, we will tell you to spend hours patching 5yo binary-only programs. This is going to make you feel super welcome, you should start to feel it any moment now. Yes, Nix was about build reproducibility, but we have a new program now. See, getting into Nix was getting a bit too easy with these new tutorials, and we already packaged most of the software, so now we're going to fix all Unicode-related bugs, in all encoding, locales and collations.

(now that I wrote this, it actually sounds quite exciting)

**Comment by purepani** [comment]
That feels a bit unnecessarily mocking. It is definitely important to try and spell words correctly, and these are modern, and important, programs that absolutely shouldn't break because there's a random non-ascii character somewhere. Even if it's a simple change technically, changing names like this are still not without their weight.
I do hope that the constitutional assembly can make these decisions feel a little less ad hoc in the future, because this did feel a bit *too* arbitrary

**Comment by djacu** [comment]
> I do hope that the constitutional assembly can make these decisions feel a little less ad hoc in the future, because this did feel a bit _too_ arbitrary

I don't see what's arbitrary about breaking user space.

> 11 years later, we still do not have the backbone to unite against this sort of issues. A real loss for the celebration of diversity and upholding our values of welcoming everyone.
> 
> This is disappointing to say the least.

I also don't see the point in moral grandstanding over the tilde on ñ. Careful placing too much social and moral weight on ñ; it might crumble into glyph dust.

We tried to have Vicuña as the spelling, found out it broke packages downstream, and changed it so it didn't. Maintainers of the affected packages are aware that unicode breaks their packages and will do their best to fix it given time. We don't need to strong arm them into fixing it _right now_ because we have to have our ñ. Keeping the ñ will do as much diversity and inclusivity as pepsi and Kendal Jenner did for racism, and be just as embarrassing.

**Comment by pluiedev** [comment]
I think, putting the inclusivity argument aside, there were predominantly two main viewpoints in this discussion so far:

a) The view that .NET should have supported Unicode in their HTTP headers handling long before this happened, and
b) The view that NixOS should switch the spelling to stop the current breakage for people who depend on .NET-based software.

These two views actually don't conflict with each other. We can accept that, yeah, it's 2024, how the heck is .NET the only one still stuck with US-ASCII only as of today, but we can also accept that changing it to a plain n is the solution that we need for now, while .NET rolls out the fixes. 

If we can agree on this shared view, then there's no real need to keep this going — the debate's been had, the decision's done, we could revisit this once Vicuña's actually due for release.

As for my thoughts on the inclusivity argument... I wouldn't elevate it that far. My problem is more with people justifying the change by saying the English spelling is right (it is not), and with people who say we should just stick to plain ASCII like everyone else does because we are already used to normalizing words into ASCII (inertia). See my comment above for my reasoning against that.

I don't think anyone's *actually* insulted because the funny little tilde on the n was removed due to technical reasons, which is why while I find the mockery to be in particularly poor taste, I think there's a point to be had in not attaching that much symbolic value on that one glyph.

**Comment by purepani** [comment]
The reasons didn't feel arbitrary; I meant moreso the process. This issue was opened, basically immedietly closed without much discussion, then opened again some time later with no apparent impetus, and then merged relatively quickly.
Maybe arbitrary isn't the best word; maybe random? The entire process of it getting merged happened while I was asleep, and given some of the responses, it wasn't particularly obvious whether the change was temporary, or permenant, etc. Yes it was stated that it would be changed back once fixed, but the comments on this pull request are plentiful so it would be easy to miss. 

It also didn't feel like the all of the dotnet tooling breaking made as big of a deal as maybe it should have. The PR was initially closed on the grounds that there wasn't a showstopper, which the dotnet tooling being unusable definitely would qualify for at least imo, so there seemed to be at least some misunderstanding there, and after that it took quite a while for someone to reopen the PR. While the change was made relatively quickly, and despite somewhat closely following this on github, I can't tell you where this issue fell through, and how it got picked up again, and I also can't quite tell how I could have raised importance for the issue aside from being loud and pinging semi-random people. Hence the feeling of it being a bit arbitrary. 

Having more triaging would likely help tremendously.

**Comment by Mic92** [comment]
Ok. No we are going off-topic here. For people interested in fixing the issue itself, you can discuss it here: https://github.com/NixOS/nixpkgs/issues/315574

**Pull request merged**